data/org.freedesktop.portal.Usb.xml

```diff
@@ -0,0 +1,222 @@
+    </method>
+
+    <!--
+        AcquireDevices:
+        @parent_window: Identifier for the application window, see :ref:`Common Conventions <window-identifiers>`.
+        @devices: Array of device identifiers, device information, and access options
+        @options: Vardict with optional further information
+        @handle: Object path for the :ref:`org.freedesktop.portal.Request` object representing this call
+
+        Acquires the given device nodes.
+
+        Each element of the @devices array contains the device ID, and the
+        following keys:
+
+        * ``writable`` (``b``)
+
+          Whether the device will be opened in read-write or read-only mode.
+          Default: False
+
+        Supported keys in the @options vardict include:
+
+        * ``handle_token`` (``s``)
+
+          A string that will be used as the last element of the @handle. Must be a valid
+          object path element. See the :ref:`org.freedesktop.portal.Request` documentation
+          for more information about the @handle.
+
+        The #org.freedesktop.portal.Request::Response signal is emitted without
+        any extra information.
+     -->
+    <method name="AcquireDevices">
+      <arg type="s" name="parent_window" direction="in"/>
+      <arg type="a(sa{sv})" name="devices" direction="in"/>
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="o" name="handle" direction="out"/>
+    </method>
+
+    <!--
+        FinishAcquireDevices:
+        @options: Vardict with optional further information
+        @results: Array of device ids, and the result of the access
+        @finished: Whether all device results were reported
+
+        Retrieves the file descriptors of the devices requested during
+        org.freedesktop.portal.Usb.AcquireDevices().
+
+        This method can only be called once, and only after calling
+        org.freedesktop.portal.Usb.AcquireDevices().
+
+        Each element of the @devices_fds array contains the device ID, and the
+        following keys:
+
+        * ``success`` (``b``)
+
+          Whether the device access was successful or not.
+
+        * ``fd`` (``h``)
+
+          The file descriptor representing the device. Only present if this
+          was a successful device access.
+
+        * ``error`` (``s``)
+
+          Error message describing why accessing the device was not
+          successful. Only present if this was an failed device access.
+
+        There are no supported keys in the @options vardict.
+     -->
+    <method name="FinishAcquireDevices">
+      <arg type="a{sv}" name="options" direction="in"/>
+      <arg type="a(sa{sv})" name="results" direction="out"/>
+      <arg type="b" name="finished" direction="out"/>
+    </method>
+
+    <!--
+        ReleaseDevices:
+        @devices: Array of device identifiers
+        @options: Vardict with optional further information
+
+        Releases previously acquired devices.
```

Comment:
What does it mean to release a device? Is the fd still usable?

Reply:
It will close the file descriptor when we release the "owned device" object. I haven't checked how libusb, for example, reacts to this.

Reply:
Well, it closes the fd on the portal side which is good because we need to do that to not leak things, but it doesn't close the fd on the client side. In fact, you cannot close the fd on the client side from the portal side. The only thing we can do is revoke file descriptors but that currently only works for evdev and hidraw and not for usb. So as long as we don't get kernel support for that, clients will be able to continue to use the fd that they acquired even if the connection is closed. We should however say something in the interface description here which will allow us to revoke those USB devices if the connection gets closed, or if the user wants to revoke access manually, when we get USB revoke from the kernel.